# Assignment: Segmenting and Clustering Neighborhoods in Toronto
by Uygar Hizal

<div class="alert alert-success" role="alert">
<h2 class="alert-heading">Table of Contents </h2>
 
    
  <h3> <a href="#item1" class="alert-link">  1. Wikipedia Scraper with Results stored into a data frame (Assignment 1)</a><br> </h3>
  <h3> <a href="#item2" class="alert-link">  2. Adding Latitude and Longitude values to our data frame (Assignment 2)</a><br>  </h3>
  <h3> <a href="#item3" class="alert-link">  3. Explore Toronto Neighborhoods / Visualize and Cluster (Assignment 3)</a><br>  </h3>
   
  <ul> 
   <li><h4> <a href="#item3A" class="alert-link"> 3.A) Getting venues from the only "first" Neighborhood</a> <br>  </h4> </li>    
   <li><h4> <a href="#item3B" class="alert-link"> 3.B) Getting venues from the all Neighborhoods (Containing Toronto) with a Function</a><br>   </h4> </li>    
   <li><h4> <a href="#item3C" class="alert-link"> 3.C) Analyze Each Neighborhood</a> <br>   </h4>  </li>   
   <li><h4> <a href="#item3D" class="alert-link"> 3.D) Cluster Neighborhoods</a> <br>   </h4>  </li>   
   <li><h4> <a href="#item3E" class="alert-link"> 3.E) Examine Clusters</a> <br> </h4> </li> </li>
    </ul>    
  

        
</div>


In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

## <--------------------------------- Assignment 1 ---------------------------------->

## Part 1 : Wikipedia Scraper with Results stored into a data frame

#### Requirement : Use the BeautifulSoup package OR ANY OTHER WAY you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe

In [2]:
# This is my method after some search : reliable and clean. (NOT the BeautifulSoup ! )..I hope you like it too..

url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df1 = pd.read_html(url)[0]


In [3]:
df1.shape

(288, 3)

In [4]:
df1.head(20)

Postcode           Borough     Neighbourhood
0       M1A      Not assigned      Not assigned
1       M2A      Not assigned      Not assigned
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
9       M8A      Not assigned      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern
13      M2B      Not assigned      Not assigned
14      M3B        North York   Don Mills North
15      M4B         East York  Woodbine Gardens
16      M4B         East York     Parkview Hill
17      M5B  Downtown Toronto           Ryerson
18      M5B  Downtown Toronto   Garden District
19      M6B        North York         Glencairn

#### Requirement : Only process the cells that have an assigned borough.Ignore cells with a borough that is Not assigned.
to do this : we simply drop rows, with coloumn Borough == 'Not assigned'  

In [5]:
df1.drop(df1.loc[df1['Borough']=='Not assigned'].index, inplace=True)

In [6]:
df1.shape

(211, 3)

As we can see from the results of shape methods above , rows are decreased from 288 to 211. 
Not assigned Boroughs are not more in our Data Frame (df1)..

In [7]:
#Lets check our dataframe again, looks better.
df1.head(20)

Postcode           Borough     Neighbourhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern
14      M3B        North York   Don Mills North
15      M4B         East York  Woodbine Gardens
16      M4B         East York     Parkview Hill
17      M5B  Downtown Toronto           Ryerson
18      M5B  Downtown Toronto   Garden District
19      M6B        North York         Glencairn
22      M9B         Etobicoke        Cloverdale
23      M9B         Etobicoke         Islington
24      M9B         Etobicoke      Martin Grove
25      M9B         Etobicoke  Princess Gardens

#### Requirement : Combine the neighborhoods which exist more than one for a postal code into one row as comma separated :

In [8]:
df2 = df1.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(', '.join).reset_index()
df2.columns = ['Postcode', 'Borough', 'Neighborhood']
df2.head(11)

Postcode      Borough                                       Neighborhood
0       M1B  Scarborough                                     Rouge, Malvern
1       M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2       M1E  Scarborough                  Guildwood, Morningside, West Hill
3       M1G  Scarborough                                             Woburn
4       M1H  Scarborough                                          Cedarbrae
5       M1J  Scarborough                                Scarborough Village
6       M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park
7       M1L  Scarborough                    Clairlea, Golden Mile, Oakridge
8       M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West
9       M1N  Scarborough                        Birch Cliff, Cliffside West
10      M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...

#### Requirement : If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [9]:
df2.loc[df2.Neighborhood == 'Not assigned', "Neighborhood"] = df2.Borough

In [10]:
# to be sure if the required process is succesfully done ,
# We can check Queen's Park  example (row nr 85), as we know it had 'Not assigned' neighborhood from the lecturer's explanation..

df2

Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
30       M3K        North York   
31       M3L        North York   
32       M3M        North York   
33       M3N        North York   
34       M4A        North York   
35       M4B         East York   
36       M4C         East York   
37       M4E      East Toronto   
38       M4G         East York   
39       M4H         East York   
40       M4J         East York   
41       M4K      East Toronto   
42       M4L      East Toronto   
43       M4M      East Toronto   
44       M4N   Central Toronto   
45       M4P   Central Toronto   
46       M4R   Central Toronto   
47       M4S   Central Toronto   
48       M4T   Central Toronto   
49       M4V   Central Toronto   
50       M4W  Downtown Toronto   
51       M4X  Downtown Toronto   
52       M4Y  Downtown Toronto   
53       M5A  Downtown Toronto   
54       M5B  Downtown Toronto   
55       M5C  Downtown Toronto   
56       M5E  Downtown Toronto   
57       M5G  Downtown Toronto   
58       M5H  Downtown Toronto   
59       M5J  Downtown Toronto   
60       M5K  Downtown Toronto   
61       M5L  Downtown Toronto   
62       M5M        North York   
63       M5N   Central Toronto   
64       M5P   Central Toronto   
65       M5R   Central Toronto   
66       M5S  Downtown Toronto   
67       M5T  Downtown Toronto   
68       M5V  Downtown Toronto   
69       M5W  Downtown Toronto   
70       M5X  Downtown Toronto   
71       M6A        North York   
72       M6B        North York   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A         Etobicoke   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                          Neighborhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7          

#### everthing seems to be ok.. Queen's Park (row nr 85 above) had a 'Not assigned' Neigbourhood  but now it is same as Borough value, like said.

In [11]:
df2.shape

(103, 3)

<a id='item2'></a>

## <--------------------------------- Assignment 2 ---------------------------------->

# Part 2 : Adding Latitude and Longitude values to our data frame
### (Requirement : Use the Geocoder package OR THE CSV FILE to create the desired data frame)

In [12]:
# We will Read the geo data from the supplied csv  file :
df_geo = pd.read_csv('http://cocl.us/Geospatial_data')
df_geo.columns = ['Postcode', 'Latitude', 'Longitude']

And below we merge our old 'df2' data frame and newly created 'df_geo' data frames based on Postcode column..

In [13]:
df_combined = pd.merge(df2, df_geo, on=['Postcode'], how='inner')

In [14]:
df_combined.head(20)

Postcode      Borough                                       Neighborhood  \
0       M1B  Scarborough                                     Rouge, Malvern   
1       M1C  Scarborough             Highland Creek, Rouge Hill, Port Union   
2       M1E  Scarborough                  Guildwood, Morningside, West Hill   
3       M1G  Scarborough                                             Woburn   
4       M1H  Scarborough                                          Cedarbrae   
5       M1J  Scarborough                                Scarborough Village   
6       M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park   
7       M1L  Scarborough                    Clairlea, Golden Mile, Oakridge   
8       M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West   
9       M1N  Scarborough                        Birch Cliff, Cliffside West   
10      M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...   
11      M1R  Scarborough                                  Maryvale, Wexford   
12      M1S  Scarborough                                          Agincourt   
13      M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter   
14      M1V  Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...   
15      M1W  Scarborough                                    L'Amoreaux West   
16      M1X  Scarborough                                        Upper Rouge   
17      M2H   North York                                  Hillcrest Village   
18      M2J   North York                       Fairview, Henry Farm, Oriole   
19      M2K   North York                                    Bayview Village   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577  
15  43.799525 -79.318389  
16  43.836125 -79.205636  
17  43.803762 -79.363452  
18  43.778517 -79.346556  
19  43.786947 -79.385975

In [15]:
df_combined.shape

(103, 5)

<a id='item3'></a>

## <--------------------------------- Assignment 3 ---------------------------------->

# Part 3 : Explore Toronto Neighborhoods / Visualize and Cluster

In [16]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_combined['Borough'].unique()),
        df_combined.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


In [17]:
address = 'Toronto'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [18]:
neighborhoods = df_combined[['Borough', 'Neighborhood', 'Latitude', 'Longitude']].copy()

Lets Create a new data frame  : neighborhoods with only desired 4 columns : 

In [19]:
neighborhoods.head()

Borough                            Neighborhood   Latitude  Longitude
0  Scarborough                          Rouge, Malvern  43.806686 -79.194353
1  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497
2  Scarborough       Guildwood, Morningside, West Hill  43.763573 -79.188711
3  Scarborough                                  Woburn  43.770992 -79.216917
4  Scarborough                               Cedarbrae  43.773136 -79.239476

In [20]:
# create map of Toronto using latitude and longitude values for ALL 103 Neighborhoods ! 
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

The Map shows all the neighborhoods very good , however, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods whose boroughs that contain the word Toronto  . So let's slice the original data frame and create a new data frame of the boroughs containing 'toronto' or 'Toronto' data. So we will search 'oronto' string to eliminate capital letter :

![Toronto](https://raw.githubusercontent.com/2spring/Assignment-Segmenting-and-Clustering-Neighborhoods-in-Toronto/master/toronto3.png)

### Please pay attention From here we will work only with simplified data frame defined as 'toronto_data' 
### it contains  only the neighborhoods in the name Toronto string exist (38 pcs),
### Process is same if we work with all neighborhoods 
### Reason for that : To simplify rendered tables here and cleanness of the notebook. (like lecturer also made in New York lab (-->manhatttan) !)

In [21]:
toronto_data = neighborhoods[neighborhoods['Borough'].str.contains('oronto')].reset_index(drop=True)
toronto_data



Borough                                       Neighborhood  \
0       East Toronto                                        The Beaches   
1       East Toronto                       The Danforth West, Riverdale   
2       East Toronto                     The Beaches West, India Bazaar   
3       East Toronto                                    Studio District   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                                   Davisville North   
6    Central Toronto                                 North Toronto West   
7    Central Toronto                                         Davisville   
8    Central Toronto                        Moore Park, Summerhill East   
9    Central Toronto  Deer Park, Forest Hill SE, Rathnelly, South Hi...   
10  Downtown Toronto                                           Rosedale   
11  Downtown Toronto                        Cabbagetown, St. James Town   
12  Downtown Toronto                               Church and Wellesley   
13  Downtown Toronto                          Harbourfront, Regent Park   
14  Downtown Toronto                           Ryerson, Garden District   
15  Downtown Toronto                                     St. James Town   
16  Downtown Toronto                                        Berczy Park   
17  Downtown Toronto                                 Central Bay Street   
18  Downtown Toronto                           Adelaide, King, Richmond   
19  Downtown Toronto  Harbourfront East, Toronto Islands, Union Station   
20  Downtown Toronto           Design Exchange, Toronto Dominion Centre   
21  Downtown Toronto                     Commerce Court, Victoria Hotel   
22   Central Toronto                                           Roselawn   
23   Central Toronto                Forest Hill North, Forest Hill West   
24   Central Toronto                The Annex, North Midtown, Yorkville   
25  Downtown Toronto                     Harbord, University of Toronto   
26  Downtown Toronto          Chinatown, Grange Park, Kensington Market   
27  Downtown Toronto  CN Tower, Bathurst Quay, Island airport, Harbo...   
28  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
29  Downtown Toronto             First Canadian Place, Underground city   
30  Downtown Toronto                                           Christie   
31      West Toronto                       Dovercourt Village, Dufferin   
32      West Toronto                           Little Portugal, Trinity   
33      West Toronto       Brockton, Exhibition Place, Parkdale Village   
34      West Toronto                      High Park, The Junction South   
35      West Toronto                             Parkdale, Roncesvalles   
36      West Toronto                                 Runnymede, Swansea   
37      East Toronto  Business Reply Mail Processing Centre 969 Eastern   

     Latitude  Longitude  
0   43.676357 -79.293031  
1   43.679557 -79.352188  
2   43.668999 -79.315572  
3   43.659526 -79.340923  
4   43.728020 -79.388790  
5   43.712751 -79.390197  
6   43.715383 -79.405678  
7   43.704324 -79.388790  
8   43.689574 -79.383160  
9   43.686412 -79.400049  
10  43.679563 -79.377529  
11  43.667967 -79.367675  
12  43.665860 -79.383160  
13  43.654260 -79.360636  
14  43.657162 -79.378937  
15  43.651494 -79.375418  
16  43.644771 -79.373306  
17  43.657952 -79.387383  
18  43.650571 -79.384568  
19  43.640816 -79.381752  
20  43.647177 -79.381576  
21  43.648198 -79.379817  
22  43.711695 -79.416936  
23  43.696948 -79.411307  
24  43.672710 -79.405678  
25  43.662696 -79.400049  
26  43.653206 -79.400049  
27  43.628947 -79.394420  
28  43.646435 -79.374846  
29  43.648429 -79.382280  
30  43.669542 -79.422564  
31  43.669005 -79.442259  
32  43.647927 -79.419750  
33  43.636847 -79.428191  
34  43.661608 -79.464763  
35  43.648960 -79.456325  
36  43.651571 -79.484450  
37  43.662744 -79.321558

In [22]:
# RE_create map of Toronto using latitude and longitude values of only 38 Boroughs :
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Lets Create  the same Toronto map, but this time only 38 Boroughs which contains string 'Toronto'..
#### Our Map will be Cleaner  : 

![Toronto](https://raw.githubusercontent.com/2spring/Assignment-Segmenting-and-Clustering-Neighborhoods-in-Toronto/master/toronto2.png)

In [23]:
CLIENT_ID = 'M5YZDFNSYT5J1ZOI5QCPZWNIPU4L15O5YNRX2CF2MNZEKTTU' # your Foursquare ID
CLIENT_SECRET = '52ZRKKZKJ2EJM1BQECTJZLNDPOCFYEE55TYI0FA3ZFZXAP0J' # your Foursquare Secret
VERSION = '20140517' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: M5YZDFNSYT5J1ZOI5QCPZWNIPU4L15O5YNRX2CF2MNZEKTTU
CLIENT_SECRET:52ZRKKZKJ2EJM1BQECTJZLNDPOCFYEE55TYI0FA3ZFZXAP0J


## 3.A) Getting venues from the only "first" Neighborhood :

<a id='item3A'></a>

In [24]:
toronto_data.loc[0, 'Neighborhood']

'The Beaches'

In [25]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


In [26]:
LIMIT = 100
radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=M5YZDFNSYT5J1ZOI5QCPZWNIPU4L15O5YNRX2CF2MNZEKTTU&client_secret=52ZRKKZKJ2EJM1BQECTJZLNDPOCFYEE55TYI0FA3ZFZXAP0J&v=20140517&ll=43.67635739999999,-79.2930312&radius=500&limit=100'

In [27]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5cfd88891ed21914be92b9ec'},
 'response': {'headerLocation': 'The Beaches',
  'headerFullLocation': 'The Beaches, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.680857404499996,
    'lng': -79.28682091449052},
   'sw': {'lat': 43.67185739549999, 'lng': -79.29924148550948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ad4c062f964a52011f820e3',
       'name': 'The Big Carrot Natural Food Market',
       'contact': {},
       'location': {'address': '125 Southwood Dr',
        'lat': 43.678879,
        'lng': -79.297734,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.678879,
          'lng': -79.297734}],
        'distance': 471,
        'posta

In [28]:

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [29]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
1                 Grover Pub and Grub                Pub  43.679181 -79.297215
2                   Glen Manor Ravine              Trail  43.676821 -79.293942
3                       Upper Beaches       Neighborhood  43.680563 -79.292869

In [30]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

4 venues were returned by Foursquare.


## 3.B) Getting venues from the all Neighborhoods (Containing Toronto) with a Function :
#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

<a id='item3B'></a>

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Let's call the function :

In [32]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
The Annex, North Midtown, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

#### Let's check the size of the resulting dataframe

In [33]:
print(toronto_venues.shape)
toronto_venues.head()

(1695, 7)


Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031  The Big Carrot Natural Food Market       43.678879   
1              -79.293031                 Grover Pub and Grub       43.679181   
2              -79.293031                   Glen Manor Ravine       43.676821   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                            Pantheon       43.677621   

   Venue Longitude     Venue Category  
0       -79.297734  Health Food Store  
1       -79.297215                Pub  
2       -79.293942              Trail  
3       -79.292869       Neighborhood  
4       -79.351434   Greek Restaurant

#### Let's check how many venues were returned for each neighborhood

In [34]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            55   
Brockton, Exhibition Place, Parkdale Village                           21   
Business Reply Mail Processing Centre 969 Eastern                      19   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            46   
Central Bay Street                                                     88   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   87   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             38   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           20   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         34   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              48   
High Park, The Junction South                                          23   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               62   
Moore Park, Summerhill East                                             2   
North Toronto West                                                     16   
Parkdale, Roncesvalles                                                 15   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     38   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        38   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         18   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             55   
Brockton, Exhibition Place, Parkdale Village                            21   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             46   
Central Bay Street                                                      88   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                15   
Church and Wellesley                                                    87   
Commerce Court, Victoria Hotel     

#### Let's find out how many unique categories can be curated from all the returned venues

In [35]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 237 uniques categories.


## 3.C) Analyze Each Neighborhood

<a id='item3C'></a>

In [36]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Auto Workshop  \
0           0                    0                 0              0   
1           0                    0                 0              0   
2           0                    0                 0              0   
3           0                    0                 0              0   
4           0                    0                 0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0                     0               0        0                0   
1                     0               0        0                0   
2                     0               0        0                0   
3                     0               0        0                0   
4                     0               0        0                0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant  Cheese Shop  \
0                          0             0                     0            0   
1                          0             0    

And let's examine the new dataframe size.

In [37]:
toronto_onehot.shape

(1695, 237)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [38]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.047619   
3   Business Reply Mail Processing Centre 969 Eastern     0.052632   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.011364   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011494   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.000000   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.020833   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.016129   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.062500   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.026316   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011494           0.011494  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.0

#### Let's confirm the new size

In [39]:
toronto_grouped.shape

(38, 237)

#### Let's print each neighborhood along with the top 5 most common venues

In [40]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2           Steakhouse  0.04
3                  Bar  0.04
4  American Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2              Bakery  0.04
3  Italian Restaurant  0.04
4      Farmers Market  0.04


----Brockton, Exhibition Place, Parkdale Village----
            venue  freq
0  Breakfast Spot  0.10
1            Café  0.10
2     Coffee Shop  0.10
3     Yoga Studio  0.05
4             Gym  0.05


----Business Reply Mail Processing Centre 969 Eastern----
              venue  freq
0       Yoga Studio  0.05
1     Auto Workshop  0.05
2              Park  0.05
3        Comic Shop  0.05
4  Recording Studio  0.05


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  0.14
1   Airport Service

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [41]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café   American Restaurant       Thai Restaurant   
1          Cocktail Bar    Seafood Restaurant                  Café   
2                  Café        Breakfast Spot           Yoga Studio   
3         Auto Workshop           Pizza Place  Gym / Fitness Center   
4       Airport Service      Airport Terminal         Boat or Ferry   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse                   Bar          Burger Joint   
1           Cheese Shop        Farmers Market              Beer Bar   
2    Italian Restaurant             Pet Store          Climbing Gym   
3      Recording Studio            Restaurant               Butcher   
4       Harbor / Marina              Boutique               Airport   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel                Bakery         Cosmetics Shop  
1            Steakhouse                Bakery     Italian Restaurant  
2            Restaurant  Caribbean Restaurant          Burrito Place  
3         Burrito Place               Brewery             Skate Park  
4    Airport Food Court          Airport Gate                  Plane

## 3.D) Cluster Neighborhoods

<a id='item3D'></a>

Run *k*-means to cluster the neighborhood into 5 clusters.

In [43]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [44]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Borough                    Neighborhood   Latitude  Longitude  \
0     East Toronto                     The Beaches  43.676357 -79.293031   
1     East Toronto    The Danforth West, Riverdale  43.679557 -79.352188   
2     East Toronto  The Beaches West, India Bazaar  43.668999 -79.315572   
3     East Toronto                 Studio District  43.659526 -79.340923   
4  Central Toronto                   Lawrence Park  43.728020 -79.388790   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0                 Trail                   Pub   
1               0      Greek Restaurant           Coffee Shop   
2               0                  Park           Pizza Place   
3               0                  Café           Coffee Shop   
4               4                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0     Health Food Store               Dog Run        Doner Restaurant   
1        Ice Cream Shop    Italian Restaurant  Furniture / Home Store   
2      Sushi Restaurant        Ice Cream Shop      Italian Restaurant   
3             Gastropub    Italian Restaurant                  Bakery   
4              Bus Line         Women's Store                 Dog Run   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0            Donut Shop   Dumpling Restaurant  Eastern European Restaurant   
1           Yoga Studio           Pizza Place                      Brewery   
2     Fish & Chips Shop          Liquor Store         Fast Food Restaurant   
3   American Restaurant           Yoga Studio                          Gym   
4  Fast Food Restaurant        Farmers Market           Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0     Electronics Store          Women's Store  
1       Bubble Tea Shop                   Café  
2         Movie Theater              Pet Store  
3               Brewery     Seafood Restaurant  
4           Event Space   Ethiopian Restaurant

Finally, let's visualize the resulting clusters

In [45]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 3.E) Examine Clusters

![Toronto](https://raw.githubusercontent.com/2spring/Assignment-Segmenting-and-Clustering-Neighborhoods-in-Toronto/master/toronto1.png)

<a id='item3E'></a>

#### Cluster 1

In [46]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                                         The Beaches                 Trail   
1                        The Danforth West, Riverdale      Greek Restaurant   
2                      The Beaches West, India Bazaar                  Park   
3                                     Studio District                  Café   
5                                    Davisville North                  Park   
6                                  North Toronto West           Coffee Shop   
7                                          Davisville           Pizza Place   
9   Deer Park, Forest Hill SE, Rathnelly, South Hi...                   Pub   
11                        Cabbagetown, St. James Town           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
13                          Harbourfront, Regent Park           Coffee Shop   
14                           Ryerson, Garden District           Coffee Shop   
15                                     St. James Town           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
18                           Adelaide, King, Richmond           Coffee Shop   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20           Design Exchange, Toronto Dominion Centre           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
24                The Annex, North Midtown, Yorkville                  Café   
25                     Harbord, University of Toronto                  Café   
26          Chinatown, Grange Park, Kensington Market                  Café   
27  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
28                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
29             First Canadian Place, Underground city                  Café   
30                                           Christie                  Café   
31                       Dovercourt Village, Dufferin                Bakery   
32                           Little Portugal, Trinity                   Bar   
33       Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
34                      High Park, The Junction South    Mexican Restaurant   
35                             Parkdale, Roncesvalles        Breakfast Spot   
36                                 Runnymede, Swansea           Pizza Place   
37  Business Reply Mail Processing Centre 969 Eastern           Yoga Studio   

            2nd Most Common Venue 3rd Most Common Venue  \
0                             Pub     Health Food Store   
1                     Coffee Shop        Ice Cream Shop   
2                     Pizza Place      Sushi Restaurant   
3                     Coffee Shop             Gastropub   
5                  Breakfast Spot     Food & Drink Shop   
6                     Yoga Studio            Bagel Shop   
7                  Sandwich Place          Dessert Shop   
9                     Coffee Shop    Light Rail Station   
11                     Restaurant                  Park   
12            Japanese Restaurant      Sushi Restaurant   
13                            Pub                Bakery   
14                 Clothing Store                  Café   
15                           Café                 Hotel   
16                   Cocktail Bar    Seafood Restaurant   
17             Italian Restaurant                  Café   
18                           Café   American Restaurant   
19                       Aquarium                 Hotel   
20                           Café                 Hotel   
21                           Café                 Hotel   
24                 Sandwich Place           Coffee Shop   
25                         Bakery                   Bar   
26  Vegetarian / Vegan Restaurant    Mexican Restaurant 

#### Cluster 2

In [47]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
8  Moore Park, Summerhill East            Playground          Tennis Court   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
8                 Diner   Filipino Restaurant  Fast Food Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
8        Farmers Market    Falafel Restaurant           Event Space   

  9th Most Common Venue 10th Most Common Venue  
8  Ethiopian Restaurant      Electronics Store

#### Cluster 3

In [48]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
10                             Rosedale                  Park   
23  Forest Hill North, Forest Hill West                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
10            Playground                 Trail                 Diner   
23         Jewelry Store      Sushi Restaurant                  Park   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
10  Fast Food Restaurant        Farmers Market    Falafel Restaurant   
23     Electronics Store      Doner Restaurant            Donut Shop   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
10           Event Space         Ethiopian Restaurant      Electronics Store  
23   Dumpling Restaurant  Eastern European Restaurant          Women's Store

#### Cluster 4

In [49]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
22     Roselawn        Ice Cream Shop                Garden   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
22         Women's Store        Discount Store  Fast Food Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
22        Farmers Market    Falafel Restaurant           Event Space   

   9th Most Common Venue 10th Most Common Venue  
22  Ethiopian Restaurant      Electronics Store

#### Cluster 5

In [50]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
4  Lawrence Park                  Park           Swim School   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4              Bus Line         Women's Store               Dog Run   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4  Fast Food Restaurant        Farmers Market    Falafel Restaurant   

  9th Most Common Venue 10th Most Common Venue  
4           Event Space   Ethiopian Restaurant

### Thank you for your reading! 
Uygar Hizal